### PR TITLE
Add debugging output to fallback creation

### PR DESCRIPTION
**Pull request by jeremylt**
The Operator fallback strikes again. This addes some mild refactoring and debugging output to help me chase a problem in Ratel.

### CODE REVIEW

File: include/ceed-impl.h [hunk 1]
```diff
@@ -387,4 +388,7 @@ struct CeedOperator_private {
   CeedContextFieldLabel *context_labels;
 };
 
+CEED_INTERN int CeedOperatorGetFallback(CeedOperator op,
```
Comment (jeremylt) [inline include/ceed-impl.h:391]: I don't want this called by a backend. We don't have any functions in here, but I'm thinking of moving all the fallback related functions to this header file.

**Comment by jeremylt** [comment]
Ok, we missed a case with fallback that was causing a hell of a confusing bug.

`/gpu/[cuda,hip]/ref` is the only impl of composite operators. That means that we have a `/gpu/[cuda,hip]/ref` composite operator holding `/gpu/[cuda,hip]/gen` operators. By some minor miracle, we somehow did not manage to have correctness issues swapping the interpretation of the backnd data back and forth between a `ref` and a `gen` backend object.... until my most recent work in Ratel.

This PR fixes the issue by switching the GPU backends to use the interface level impl of assembly for composite operators and only using the backend specific impl for single operators.

**Comment by jeremylt** [comment]
LV is being flakey, but relevant CUDA tests pass locally for me.

**Review by jedbrown — APPROVED** [review]
There's a lot going on here and it's hard for me to assess implications, but the tests pass, it's not publicly visible, and more lines are deleted than added so :+1: .

**Comment by nbeams** [comment]
Sorry, I wasn't fast enough to comment.  But doesn't this change mean we don't don't use GPU assembly at all for composite operators?  That can't be what we want, can it?

**Comment by jeremylt** [comment]
It should use GPU assembly on the sub-operators, though I just caught an issue with that logic.

**Comment by nbeams** [comment]
Yeah, if I do `rocprof` on t565, after this PR I get
```
Grad,72,418720,5815,30.087441266903166
Interp,70,255679,3652,18.372007297670365
OffsetNoTranspose,74,206558,2791,14.842380811064636
setValueK(double*, int, double),49,135040,2755,9.703401004687151
StridedNoTranspose,72,115520,1604,8.300776688843747
OffsetTranspose,70,112320,1604,8.070838276410402
CeedKernel_Hip_ref_Poisson2DApply,37,61760,1669,4.437811359963555
CeedKernel_Hip_ref_MassApply,36,59360,1648,4.265357550638546
Weight,2,9120,4560,0.6553244754350327
CeedKernel_Hip_ref_Poisson2DBuild,1,7040,7040,0.5058645073533585
CeedKernel_Hip_ref_Mass2DBuild,1,6400,6400,0.4598768248666896
StridedTranspose,2,4160,2080,0.2989199361633482
```
-- `linearAssemble` is missing.  If I go back before this PR, I see
```
"Name","Calls","TotalDurationNs","AverageNs","Percentage"
Grad,72,418560,5813,29.68346431899512
Interp,70,253120,3616,17.950780027771515
OffsetNoTranspose,74,205280,2774,14.558059908742637
setValueK(double*, int, double),49,136959,2795,9.712866947785868
StridedNoTranspose,72,116960,1624,8.294576612073943
OffsetTranspose,70,110559,1579,7.8406300928033765
CeedKernel_Hip_ref_Poisson2DApply,37,63680,1721,4.5160622320183705
CeedKernel_Hip_ref_MassApply,36,57920,1608,4.107574190931282
linearAssemble,2,19680,9840,1.39566747371422
Weight,2,9280,4640,0.6581196217514208
CeedKernel_Hip_ref_Poisson2DBuild,1,7200,7200,0.510610051358861
CeedKernel_Hip_ref_Mass2DBuild,1,6560,6560,0.46522249123807335
StridedTranspose,2,4320,2160,0.3063660308153166
```

**Pull request merged**